Convert Android-Jobs and NetworkReceiver to WorkManager

jd-alexander opened the pull request:
Closes #2499
#### What has been done to verify that this works as intended?
- [x] `AutoSendWorker` gets executed immediately when a form is finalized and if there is no network connectivity it's executed when it becomes available.
- [x] `AutoSendWorker` the job gets marked as failed when there is a failure - Having issues doing this since the `listeners` are callbacks and the code would have to wait until these listeners are executed. This can be done using a `CountDownLatch`. Currently, the job just fails silently. An approach could be to enqueue the job when the upload or authorization fails.
- [x] `AutoSendWorker` remains the only job in the queue when another is added due to a check being done every time a new job is being added to keep the current job instead of adding a new job since they both do the same thing.

- [x] `SmsSenderWorker` sends SMS submissions when it's being from the SMS transport.
- [x] `SmsSenderWorker` only gets executed when SMS submissions are made - There is a weird bug with the `WorkManager` that's causing this behaviour when checking for a `Worker` status. I have spent some time investigating but haven't figured out a solution as yet.

- Check to see if the `ServerPollingWorker` is polling similarly to the `Job` implementation that existed previously. - Will be done by QA.
- [x] Verify that the `pollServerIfNeeded` functionality is still required.

#### Why is this the best possible solution? Were any other approaches considered?
This is the best solution because the `android-job` library will be depreciated soon and the author is using `WorkManager` underneath in it's next version so having it as a dependency is counterproductive.

#### Are there any risks to merging this code? If so, what are they?
Yes, in a sense that all the functionality needs to verified thoroughly that they are still working similarly to the previous job queue implementation.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lintDebug spotbugsDebug` and confirmed all checks still pass.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

jd-alexander committed:
Converted all jobs to WorkManager workers.

All jobs were renamed to workers to match the WorkManager naming convention. The android job library was removed from the codebase along with the constructs associated with it. The worker approach was then implemented to replicate the behaviour that was being accomplished by the android job.

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into workmanager_refactor

# Resolved conflicts:
#	build.gradle

jd-alexander committed:
Resolved a style issue.

jd-alexander committed:
Corrected extras variable name and removed constraints from Sms worker.

An error was being caused due to the wrong variable name being used to access the Worker's InputData so I referenced the correct one in the SmsSenderWorker. Also it was realized that constraints were being supplied to the WorkManager that stated that SMS can only be sent when network access is present. Removed that as well.

jd-alexander committed:
Removed NetworkReceiver and replaced it with AutoSendWorker

- All jobs enqueued are unique so only one job can be on the queue at a time.

- Network constraints mimic the NetworkReceiver behaviour.

- the WorkManager is queried to check the current status of the job to see if it's running.

jd-alexander committed:
AutoSendWorker waits on results from the respective InstanceUploaders

This involved utilizing a CountDownLatch that allows us to block the thread until a callback has returned a result which would trigger the latch to countdown thus continuing the thread's execution.

jd-alexander commented:
Will be modifying that last commit as it's malformed due to an incorrect merge.

jd-alexander committed:
Merged master and resolved conflicts

jd-alexander committed:
Fixed enqueue based issues

jd-alexander commented on collect_app/src/main/java/org/odk/collect/android/jobs/AutoSendWorker.java:
```diff
@@ -1,81 +1,78 @@
-                uploadForms(context, isFormAutoSendOptionEnabled(currentNetworkInfo));
-            }
+        try {
+            countDownLatch.await(2, TimeUnit.MINUTES);
```
So I just put a random value here, where I am saying if the app doesn't get a result from the `InstanceUploader` after two minutes then it should simply just return a failure. What I want help with is selecting a realistic time that should be used based on how long forms have taken to upload in the past due to size and connectivity and/or if the `InstanceUploaer` has any time out mechanism itself that would ensure this condition is never met.

jd-alexander replied:
I could always use the other version of the `await` method that doesn't take a time. I just wanted to put a time as a fail safe so the job can fail if necessary and hand over the result to the `WorkManager`.

lognaturel replied:
`WorkManager` can run a job immediately in addition to based on a trigger so I think the eventual goal would be to replace the existing `AsyncTask`s at some point. Does that sound right to you, @jd-alexander? Having the actual upload happen here would mean not having to coordinate with yet another thread so no countdown latch or anything like this. 

`AsyncTask`s are only supposed to be for short-running tasks, right? Now that I think more about it, I'm surprised that uploads were implemented this way and that it doesn't seem to cause problems. I suppose one of the big problems with `AsyncTask`s are the coordination with the `Activity` and since the manual upload UI is blocking and autosend isn't triggered by an activity, perhaps things running longer is ok. Anyway, I assume you decided to keep the `AsyncTask`s for now to reduce the risk of regressions and that's probably a good way to start.

What would it take to know for sure whether there are any parts of the upload process that can hang indefinitely? What would be the implications of that happening? It would be good to know whether a `Worker` eventually gets killed by the system. If there are parts of the upload process that can hang indefinitely, this is a problem with the current implementation as well, right?

What are the implications of killing the job after 2 minutes? Let's say there are lots of forms with lots of media that need to be sent. The job would fail. Then it would get re-enqueued when another form were finalized or the connectivity status changed. This could be a long time in the future so if we do decide to have a timeout I think we should send a notification to the user and let them know they can manually trigger a send (if the send screen isn't hidden in admin settings).

My current leaning would be to not have the timeout. It seems that most closely matches the current behavior and I don't think there are any terrible implications though it would be good to verify that and to at least try to verify if/where there's an opportunity for something to hang.

jd-alexander replied:
Yes, I agree with the removal of the timeout. I have been trying to reproduce some of this behaviour to verify the implications of having long running jobs but I haven't gotten anything solid as yet. I am still doing some research about how the `Workers` handle long running operations and will report once I find anything useful.

jd-alexander replied:
>   Anyway, I assume you decided to keep the AsyncTasks for now to reduce the risk of regressions and that's probably a good way to start.

Yes, that's what I was thinking. Also, the `Workers` dont have proper support for asynchronous work so that's a question I am going to pose as well to get some answers from those who are shaping the current `WorkManager` API. 

> What would it take to know for sure whether there are any parts of the upload process that can hang indefinitely?

I know upload tend to hang if extremely large files are being uploaded but normally the `HttpClient` being used takes care of that due to timeouts and size limitations. I could check with @xsteelej and just ask him if he has any ideas around this since we was in the `Http` code for a long while. 

> What would be the implications of that happening? It would be good to know whether a Worker eventually gets killed by the system. 

I posted an issue for the `PeriodicWorkRequest` behaviour that I have been trying to verify so if I get a response from the Android arch team  I can pose these questions.

> If there are parts of the upload process that can hang indefinitely, this is a problem with the current implementation as well, right?

Yes because the worker is simply starting the `ServerUploader` that's wrapped in the  `AsyncTask`.

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into workmanager_refactor

jd-alexander committed:
Returns a FAILURE when the await method of the CountDownLatch has a timeout.

jd-alexander commented on collect_app/src/main/java/org/odk/collect/android/tasks/ServerPollingWorker.java:
```diff
@@ -200,10 +211,17 @@ private boolean isDeviceOnline() {
     public static void pollServerIfNeeded() {
```
@grzesiek2010 Hey! Why is `pollServerIfNeeded` necessary in the context of this job? I was thinking to remove it since the `NetworkReceiver` is being removed. However, the `AutoSendWorker` is now the new `NetworkReceiver` so if there's a solid rationale behind this it could be tied back in.

grzesiek2010 replied:
Let's imagine a case that you selected `Every 24 hours` option to check for updates. You did that let's say on Monday at 1pm so you can expect next checks on Tuesday, Wednesday etc at the same time. But if your device has no internet connection at this particular time you lost your chance and you have to wait another 24 hours. The method you asked about is used to run the job immediately after connecting to the internet if the last attempt failed.

lognaturel commented on build.gradle:
```diff
@@ -7,8 +7,9 @@ buildscript {
-        classpath 'org.ow2.asm:asm:6.0' // https://github.com/jacoco/jacoco/issues/639#issuecomment-355424756
+        classpath 'com.google.gms:google-services:4.0.1'
+        classpath 'org.ow2.asm:asm:6.0'
+        // https://github.com/jacoco/jacoco/issues/639#issuecomment-355424756
```
This comment goes with the asm dependency so please make sure it's clearly grouped with it (either above or next to).

lognaturel commented on collect_app/build.gradle:
```diff
@@ -271,6 +271,9 @@ dependencies {
     implementation "com.trello.rxlifecycle2:rxlifecycle-android:${rootProject.rxLifecycleVersion}"
     implementation "com.trello.rxlifecycle2:rxlifecycle-android-lifecycle:${rootProject.rxLifecycleVersion}"
 
+    implementation "android.arch.work:work-runtime:${rootProject.workManagerVersion}" // use -ktx for Kotlin
```
Please remove the Kotlin comment.

Is this redundant with `implementation "android.arch.work:work-runtime:CURRENT_VERSION"` above?

jd-alexander replied:
Yes, it can be removed. Didn't notice it. Brought int he dependencies from the work manager sample repo.

lognaturel replied:
I recommend making these kinds of easy changes right away so they don't get lost!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -2830,8 +2836,18 @@ protected void onStop() {
 
-    private void sendSavedBroadcast() {
-        sendBroadcast(new Intent("org.odk.collect.android.FormSaved"));
+    void enqueueFinalizedForm() {
```
This is a misleading name because it suggests that there is a queue that the finalized form gets added to when in fact it's that an autosend job gets enqueued. Maybe `enqueueAutosendJob` or something?

It feels unusual that the job gets enqueued without a check for whether autosend is even on. I know this matches the prior structure but it feels worse now that it's "enqueue a job that may not fire for a while but that may do no work when it's triggered if auto-send is off." It used to be "trigger a send attempt on finalization, it may do no work if auto-send is off." Am I interpreting all this correctly? How much restructuring would it take to only enqueue the job if autosend is enabled (at the settings or form level)?

lognaturel replied:
It's also important that if the setting is set to Wifi only, the forms only send when wifi becomes available and the same for data.

jd-alexander replied:
Yes, I just did a check and realized that the functionality to check if auto-send is enabled was within the `NetworkReceiver` and it was moved over with the refactor. While that worked, you are right that the code can be further refactored so the check can be done before the job is added.

jd-alexander replied:
Since the `AutoSendWorker` is functioning as a job that sends all finalized form there's a function within it that checks if `auto-send` is enabled at the app level and also at the form level. To confirm are you saying that before enqueuing the job a similar check should be done externally with the form that's performing to request?

lognaturel replied:
Right now every single form finalization results in the auto send job being enqueued, correct? And then if there's connectivity, the job immediately checks if `auto-send` is enabled and completes if not. But if there's no connectivity, the job stays scheduled until connectivity becomes available and then if the check sees that `auto-send` is not enabled, it completes. Is that correct? It doesn't seem right that a job would stay enqueued if `auto-send` is off in the first place.

jd-alexander replied:
I agree. Thanks for spotting this.

lognaturel replied:
Kind of related -- where is it enforced that only one of these jobs is ever enqueued? As you said, since the job is to send all forms, I believe there should only ever be either 0 or 1 enqueued. So if a form is finalized while connectivity is off and auto-send is on, that should enqueue the work. Then if another form is finalized while connectivity is still off and auto-send is still on, nothing should happen because the work is already enqueued.

jd-alexander replied:
Yes, that's how the `WorkManager`s api is currently being utilized. Referenced that call in another comment.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java:
```diff
@@ -156,29 +162,39 @@ public void onUploadButtonsClicked(Button button) {
 
-        int checkedItemCount = getCheckedCount();
-        logger.logAction(this, "uploadButton", Integer.toString(checkedItemCount));
+            LiveData<List<WorkStatus>> statuses = WorkManager.getInstance().getStatusesForUniqueWork(AutoSendWorker.TAG);
```
I think it's worth adding a comment on this block since the return is buried and it may not be immediately clear what's going on. Something like "Disable upload button and show toast if auto-send is in progress".

lognaturel replied:
This is what it /should/ do but see below, I don't think it's actually what it does currently.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java:
```diff
@@ -156,29 +162,39 @@ public void onUploadButtonsClicked(Button button) {
+                    }
+                }
+
+                int checkedItemCount = getCheckedCount();
```
Why is this in the statuses observer?

jd-alexander replied:
This is the only way to get an idea of the jobs that are running and if so mimic the previous behaviour that the `NetworkReceiver` allowed.

jd-alexander replied:
I was trying to do it in a sync manner but was having threading related issues due to the `WorkManager` restricting runs on the main thread. I'm gonna investigate.

lognaturel replied:
From Slack: "`statuses` is a variable local to `onUploadButtonsClicked`. You put an observer on it. But then the method immediately ends. So doesn't that mean the observer may or may not execute?" I don't see how this can work consistently.

lognaturel replied:
What did you try to make it synchronous?

I believe this does what you want:
```java
            // Disable upload button and show toast if auto-send is in progress
            List<WorkStatus> statuses = WorkManager.getInstance().getStatusesForUniqueWork(AutoSendWorker.TAG).getValue();
            if (statuses != null) {
                for (WorkStatus status : statuses) {
                    if (status.getState().equals(State.RUNNING)) {
                        ToastUtils.showShortToast(R.string.send_in_progress);
                        return;
                    }
                }
            }

            int checkedItemCount = getCheckedCount();
            logger.logAction(this, "uploadButton", Integer.toString(checkedItemCount));

            if (checkedItemCount > 0) {
                // items selected
                uploadSelectedFiles(button.getId());
                setAllToCheckedState(listView, false);
                toggleButtonLabel(findViewById(R.id.toggle_button), listView);
                uploadButton.setEnabled(false);
                smsUploadButton.setEnabled(false);
            } else {
                // no items selected
                ToastUtils.showLongToast(R.string.noselect_error);
            }
```

If you wanted to actually use `LiveData`, the structure would need to change quite a bit. But in this case I don't see any advantage -- we just want to determine at a point in time when the user has clicked the button whether or not auto send is in progress. We don't care about changes over time.

jd-alexander replied:
Yes, that's true. I don't really want to use `LiveData`. Being able to simply determine if it's running or not should suffice. I am giving this approach a try.

lognaturel replied:
To provide an update here -- the `LiveData` object may not immediately have a value to get so the approach I proposed won't work. Really, the issue with the original code is that the observer runs for the entire lifetime of `InstanceUploaderList`. @jd-alexander is trying an approach that registers an observer in `onCreate` and uses it to set a flag that keeps track of whether or not an auto update is ongoing. That flag can then be read in the click handler and used to determine whether or not to start a new download.

If anyone has more experience with these frameworks and has an alternate suggestion, please jump in! [Slack thread](https://opendatakit.slack.com/archives/C35ENHA64/p1536787044000100)

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java:
```diff
@@ -156,29 +162,39 @@ public void onUploadButtonsClicked(Button button) {
+                    for (WorkStatus status : workStatuses) {
+                        if (status.getState().equals(State.RUNNING)) {
+                            ToastUtils.showShortToast(R.string.send_in_progress);
+                            return;
```
This doesn't return out of `onUploadButtonsClicked`, does it? This means an upload can be triggered while an auto-send is ongoing which is not desired.

lognaturel replied:
Ah, or is this why the actual sending is done from within the observer?

lognaturel reviewed (commented):
Not done yet but wanted to get a first batch of questions and comments out.

lognaturel commented:
Good news is that the `android-job` migration was very straightforward. The things I see remaining in addition to the unchecked items in the PR description are:
- [ ] See if check on whether auto-send is enabled can be done before enqueuing the auto-send job [here](https://github.com/opendatakit/collect/pull/2511#discussion_r217177289)
- [ ] Address upload button block issue [here](https://github.com/opendatakit/collect/pull/2511/files#r217190395)
- [ ] Address questions about `AsyncTask` [here](https://github.com/opendatakit/collect/pull/2511#discussion_r217215644)
- [ ] Make decision on whether to have a timeout on the countdown latch
- [ ] Retry form download on periodic check failure [here](https://github.com/opendatakit/collect/pull/2511#discussion_r216650638). Can you simply enqueue a job triggered on connectivity availability 
- [ ] Fix conflicts once https://github.com/opendatakit/collect/pull/2463 is merged

jd-alexander commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -2830,8 +2836,18 @@ protected void onStop() {
+                        .setConstraints(constraints)
+                        .build();
+
+        WorkManager.getInstance().beginUniqueWork(AutoSendWorker.TAG, ExistingWorkPolicy.KEEP, autoSendWork).enqueue();
```
@lognaturel Enforcement takes place right here. An unique job will be enqueued only if one doesn't exist. Thus the usage of of the `ExistingWorkPolicy` that "keeps" a job that's already enqueued.

lognaturel replied:
Perfect, thank you!

jd-alexander committed:
AutoSendWorker now returns/ends for all cases and won't be started before form is verified
                                                    Previously, some of the returns weren't counting down the CountDownLatch so even though the job would end the latch would still be awaited.
                                                    A check is done to verify that auto-send is enabled for a form before the autoSendWorker is enqueued.

jd-alexander committed:
InstanceUploaderList monitors the status of the AutoSendWorker so the UI updates accordingly.

When the worker is running a toast is shown to denote that. Once the job is complete, the observer reacts and toggles the state of the upload button.

jd-alexander committed:
Added dependency comment in the right place.

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into workmanager_refactor

# Resolved conflicts:
#	collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java
#	collect_app/src/main/java/org/odk/collect/android/jobs/AutoSendWorker.java
#	collect_app/src/main/java/org/odk/collect/android/jobs/CollectJobCreator.java
#	collect_app/src/main/java/org/odk/collect/android/services/FormDownloadService.java
#	collect_app/src/main/java/org/odk/collect/android/tasks/FormDownloadWorker.java

jd-alexander committed:
Added SMSSenderWorker to repo.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java:
```diff
@@ -2647,7 +2649,7 @@ public void savingComplete(SaveResult saveResult) {
                     formController.getTimerLogger().logTimerEvent(TimerLogger.EventTypes.FORM_EXIT, 0, null, false, true);         // Force writing of audit since we are exiting
                 }
-                enqueueFinalizedForm();
+                enqueueAutosendJob();
```
This doesn't always enqueue so `enqueueAutosendJobIfNeeded` might be more accurate. Another option is to pull out the enqueuing condition so it's really clear when jobs are and aren't enqueued.

jd-alexander replied:
Thanks. that makes sense 😄

jd-alexander replied:
I'm going to make a little PR checklist of some of these highlights for my personal reviews. Because at first addng `IfNeeded` didn't come to mind since previously the `AutoSendWorker` was being enqueued no matter what but with the changes in logic comes changes to the method's naming as well 😄

lognaturel replied:
> with the changes in logic comes changes to the method's naming as well

Yes, absolutely. It may seem nit-picky but it's the kind of thing that can cause a lot of confusion in the future!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java:
```diff
@@ -146,6 +147,23 @@ public void denied() {
         });
     }
 
+    private void isAutoSendRunning() {
```
`isAutoSendRunning` is a great name for a method that returns a boolean. For this one, maybe `startAutosendObserver` or something like that. It should be clear you're starting a process that will be ongoing during the lifetime of this activity. A comment might also help.

Did you deliberately show a toast here? I believe this means that if autosend is ongoing when the user first reaches the activity, they'll see the toast. Or if it starts while they're in this activity, they'll see it. They'll also see it when they tap the upload button. That's not necessarily a bad thing but I want to make sure it's deliberate. If it's intentional, there should be a comment to indicate this.

jd-alexander replied:
Honestly, it doesn't have to be there. I was just thinking that when the user comes unto the screen, they could get an update that background send is already sending their forms, so they don't have to do anything. The messaging would have to be changed since it is similar to the manual send button. Was thinking to do that but was just pushing the changes I already made. If it's better they try to upload then hear that it's already running similar to the previous behaviour then that's fine with me as well.

lognaturel replied:
My sense is that it's not immediately obvious to a user why they should care so I'd tend to leave it out here. That is, when they tap the upload button, the toast explains why there's no submission. If they just see the toast without any context, it's not clear what it means.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderList.java:
```diff
@@ -239,7 +254,7 @@ private void init() {
         listView.setItemsCanFocus(false);
         listView.addOnLayoutChangeListener((v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> {
-            uploadButton.setEnabled(areCheckedItems());
+            uploadButton.setEnabled(isReadyForUpload());
```
I think we had decided to leave the button enabled and show the toast as needed to match existing behavior. If the button is disabled, the toast from `onUploadButtonsClicked` is never shown, right?

jd-alexander replied:
What I thought based on the conversation we had in slack was that we would disable the button until the status for the `AutoSendWorker` has been set before enabling it. I am going to add that code to the `observer` that runs this `setEnabled` logic again if that behaviour makes sense. That's the reason for creating this method to toggle the button's state because it now depends on the `autoSendRunning` value.

lognaturel replied:
Oh, I was going too quickly and didn't realize `autoSendRunning` was a `Boolean` instead of a `boolean`. I guess I thought it was checking for `false`. Ok, this seems fine.

lognaturel commented:
Given comments [here](https://github.com/opendatakit/collect/pull/2511#discussion_r218189625), we should bump this to the next release since there are still a number of unanswered questions. I've removed the `high priority` tag but since it's high risk, I think it would be great to get it done as soon as possible so it can be thoroughly vetted.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/workers/SmsSenderWorker.java:
```diff
@@ -0,0 +1,28 @@
+/***
+ * Background job that adheres to the fire and forget architecture pattern
+ * where it's sole purpose is to send an SMS message to a destination without
+ * caring about it's response.
```
In both cases, "it's" should be "its" because they can't be replaced by "it is" and are instead possessives.

Are you sure it's fire and forget? I would expect that to mean no value at all were returned.

jd-alexander replied:
True, I was thinking of fire and forget in the context of this `Worker` not getting a response, not the entire app. The app does get a response from the `BroadcastReceiver` so if the whole fire and forget notion isn't fitting it can be removed 😅

jd-alexander replied:
> In both cases, "it's" should be "its" because they can't be replaced by "it is" and are instead possessives.

Yes, this comment was from the `SmsSenderJob` from way before I registered mentally the diff between the two. Thanks. will review the other comments as well.

lognaturel replied:
I'm not sure! It seems the worker does get a notion of success or failure back. I would expect fire and forget to mean there's something like an event fired that may trigger some behavior but whatever fired the event never hears back at all. What reference are you using for pattern names?

jd-alexander replied:
True, yes the `Worker` gets an event that the message has been sent so, to be honest, it's not required there. Some of the patterns I knew of from using message patterns when doing a bit of enterprise backend dev. I am realizing the ideas might be similar but not necessarily applicable in a mobile architecture.

jd-alexander committed:
Optimized InstanceUploaderList behaviour

Toasts are only shown when button is manually clicked.
Upload button's state is tied to auto-send observer

jd-alexander commented:
Updating the checklist!

- [x]  See if check on whether auto-send is enabled can be done before enqueuing the auto-send job [here](https://github.com/opendatakit/collect/pull/2511#discussion_r217177289)
- [x]  Address upload button block issue [here](https://github.com/opendatakit/collect/pull/2511/files#r217190395)
 - [x]  Address questions about `AsyncTask` [here](https://github.com/opendatakit/collect/pull/2511#discussion_r217215644)
 - [x]  Make decision on whether to have a timeout on the countdown latch
 - [ ]  Retry form download on periodic check failure [here](https://github.com/opendatakit/collect/pull/2511#discussion_r216650638). Can you simply enqueue a job triggered on connectivity availability
 - [x]  Fix conflicts once #2463 is merged

jd-alexander committed:
ServerPolling job now executes immediately if it missed its execution window due to lack of connectivity.

So if a job runs after specified period and there's no network connectivity for it then a job is enqueued to execute immediately.

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into workmanager_refactor

# Resolved conflicts:
#	collect_app/src/main/java/org/odk/collect/android/preferences/FormManagementPreferences.java

jd-alexander commented:
So the `ServerPollingJob` is ready for testing. I'm still in talks with one of the engineers working on the `WorkManager` to get some more details on how network constraints are handled with the `PeriodicWorkRequest`

lognaturel commented:
> get some more details on how network constraints are handled with the `PeriodicWorkRequest`

Looks like good news on that front at https://issuetracker.google.com/issues/115849161 👍 

This is going to be a tricky one to verify. All behavior around auto-sending (especially how it interacts with manual sends -- they should be disabled while autosend is ongoing), polling for form updates, automatically downloading form updates, sending SMS needs to be revisited. In particular, we need to watch for bad behavior when Internet connectivity comes in and out, when network tasks take a really long time (e.g. uploading or downloading large media files) and other special cases like that.

lognaturel commented:
Holding off on the `needs testing` tag because the feedback from the conversation with Google hasn't been incorporated yet ([Slack conversation](https://opendatakit.slack.com/archives/C35ENHA64/p1537466608000100)).

@jd-alexander I think it would also be good to update to [alpha09](https://developer.android.com/jetpack/docs/release-notes) before QA. It looks like it may not play nice with all other Google dependencies (https://issuetracker.google.com/issues/116193487) so would be good to try to run it first.

grzesiek2010 committed:
App crashes on Selfie video widget if microphone permissions are not set (#2566)

grzesiek2010 committed:
Suuport readOnly option in the ArbitraryWidget (#2561)

grzesiek2010 committed:
Support read only for rank (#2555)

grzesiek2010 committed:
Support readOnly option in SelectMap widgets (#2559)

grzesiek2010 committed:
Support rank questions in field-list (#2556)

lognaturel committed:
Update September translations

lognaturel committed:
Remove Firebase performance monitoring (#2546)

grzesiek2010 committed:
Fix attempt to verify if null dialog is showing (#2587)

grzesiek2010 committed:
Fixed android.view.WindowManager: Unable to add window -- token android.os.BinderProxy@a68b01e is not valid (#2578)

grzesiek2010 committed:
Fixed java.lang.RuntimeException: Can't create handler inside thread that has not called Looper.prepare() (#2580)

grzesiek2010 committed:
Fixed Fatal Exception: java.lang.StringIndexOutOfBoundsException: length=0; index=0 (#2582)

grzesiek2010 committed:
Fixed: attempt to re-open an already-closed object (#2584)

lognaturel committed:
Save HttpContext between requests (#2594)

lognaturel committed:
Strip out the search() appearance/function (#2591)

jd-alexander committed:
Updated WorkManager to the latest alpha version and resolved dependency conflicts.

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into workmanager_refactor

jd-alexander commented:
@lognaturel Updated to alpha and merged with master. I am just gonna do a final run through then comment here once I'm done with that before we start testing.

jd-alexander committed:
Removed manual server polling since the WorkManager does this by default.

Currently the ServerPollingWorker would enqueue a one time job if the device doesn't have internet so that would execute when connectivity is restored. An engineer from the arch components team stated that this isn't necessary because when the worker misses it's execution window due to a missing constraint it runs again after the flex period so this was removed and the network constraint added to the PeriodicWorkRequest.

jd-alexander commented:
@grzesiek2010 I removed the manual polling of the server per[ this conversation with the WorkManager team ](https://issuetracker.google.com/issues/115849161) They are basically saying that if periodic work takes place and the constraints aren't met (eg. network connectivity) then the worker runs immediately when the constraints are met (eg. when internet is restored). 

What's left now is the verification that this works well. I guess that can be done by QA?

grzesiek2010 commented:
> I guess that can be done by QA?

sure.

jd-alexander committed:
Minor style changes

jd-alexander committed:
Merge remote-tracking branch 'upstream/master' into workmanager_refactor

jd-alexander commented:
Ready for testing! @grzesiek2010 can you add it for me? Just made a minor fix

lognaturel committed:
Fix NPE when checking current connection type

lognaturel commented:
The highest risk here is around auto-send. For the other functionality, we can always go back to `android-job` if there are issues but it's auto-send that we don't currently have a good alternative for and that needs to be changed to target API 24.

As-is, Android 5.1 and Android 6.0 crashed when finalizing a form while Wifi is off:
```
    java.lang.NullPointerException: Attempt to invoke virtual method 'int android.net.NetworkInfo.getType()' on a null object reference
        at org.odk.collect.android.workers.AutoSendWorker.isFormAutoSendOptionEnabledForNetwork(AutoSendWorker.java:89)
```

I pushed a null check for that but it probably means @jd-alexander did not try to see if auto-send works. I think we had a misunderstanding and that `enqueueAutosendJobIfNeeded` may not make sense anymore. I believe it only enqueues work if the network connection is available **at the time of finalization**. The check it should do is on whether one of the autosend options is enabled **at all** since we want to enqueue the job for later. Then, when a connectivity change happens, the job should check to see if the connectivity type (data vs wifi) matches the autosend rule. `isFormAutoSendOptionEnabledForNetwork` is misnamed -- I believe it currently does something more like `isNetworkForAutosendCurrentlyAvailable`. That means that if the network is unavailable, no autosend task is enqueued.

All that said, a QA pass will need to wait for that to be addressed. In the mean time, @kkrawczyk123 there is one thing that I would really appreciate a check on and that is whether you are able to get autosend to work when a connection is available on Android < 5.0. On Android 5.1 and Android 6.0, if I finalize a filled form when WiFi is available, it successfully sends immediately. On Android 4.1.1, nothing happens.

jd-alexander commented:
>  I think we had a misunderstanding and that `enqueueAutosendJobIfNeeded` may not make sense anymore. I believe it only enqueues work if the network connection is available **at the time of finalization**. The check it should do is on whether one of the autosend options is enabled **at all** since we want to enqueue the job for later. Then, when a connectivity change happens, the job should check to see if the connectivity type (data vs wifi) matches the autosend rule. `isFormAutoSendOptionEnabledForNetwork` is misnamed -- I believe it currently does something more like `isNetworkForAutosendCurrentlyAvailable`. That means that if the network is unavailable, no autosend task is enqueued.

Okay, understood. The reason for keeping the `isFormAutoSendOptionEnabledForNetwork` was due to the fact that the current `auto-send` setting is included there but  I totally didn't realize that the network check was being included. That was an oversight on my part.

I was thinking to simply access the `SharedPreferences` autosend value and pass that result to the `isFormAutoSendEnabled`function.

lognaturel commented:
Ok! I have something that I have verified works on Android 5.1 and 7.0 in these conditions:
- finalized form is saved when there is no wifi
- wifi resumes when the app is in the foreground

It's at https://github.com/lognaturel/collect/commits/collect-2511-ln. @kkrawczyk123, please use that branch. The first priority is still to see if you can get any auto-send behavior working at all on 4.x. I still can't on 4.1. It’s possible 4.4 is different. If 4.4 works in the most basic case of saving a form while WiFi is on and autosend is on, please try some of the connectivity changes I describe below on 4.4 as well. 

The next priority is to see whether this works more broadly on Android 5.0+. Does autosend get triggered when Collect is in the background? After a reboot? After a force kill of Collect? Any other creative conditions you can think of? For now, please use Android 6 because I don't have a device that runs it and it is a popular version (21% of users). I expect other versions 5.0+ will work the same way so let's get that information first to see whether this is a direction worth pursuing and then we can do a deeper pass after some cleanup. 

Unlike the prior functionality, there can be a delay between when the connection becomes available and when the auto-send occurs. I think that for this initial verification we can wait for 10 minutes and if auto-send doesn't run, consider it a failure.

@jd-alexander Since this is blocking for the target API increase, I can take it from here to minimize the back-and-forth and will create a new PR tomorrow. Thanks for your work on it so far. If you have any insights on expected behavior on 4.x from your research or your own manual verification, that would be much appreciated.

kkrawczyk123 commented:
> All that said, a QA pass will need to wait for that to be addressed. In the mean time, @kkrawczyk123 there is one thing that I would really appreciate a check on and that is whether you are able to get autosend to work when a connection is available on Android < 5.0. On Android 5.1 and Android 6.0, if I finalize a filled form when WiFi is available, it successfully sends immediately. On Android 4.1.1, nothing happens.

I have checked and autosend works for me for all Androids on which I was testing (4.1, 4.2, 4.4, 5.1, 6.0 and 8.1) in the same way. 
**Case which works is:** autosend wifi only is on, wifi is on, I finalized form when I was online - right after finalize it it was imediately sent.
**Case which doesn not work:** autosend wifi only is on, wifi is off, I finalize form when I am offline and then I turn on wifi - nothing happens for all devices, there is no failed sending status, just finalized like it won't even try to send it. 
The exact behavior was with cellular connection try. 

> Ok! I have something that I have verified works on Android 5.1 and 7.0 in these conditions: finalized form is saved when there is no wifi- wifi resumes when the app is in the foreground
It's at https://github.com/lognaturel/collect/commits/collect-2511-ln. @kkrawczyk123, please use that branch. The first priority is still to see if you can get any auto-send behavior working at all on 4.x. I still can't.

I used that branch and I was able to autosend a form no matter if I finalized form when I was online or not - both cases succeeded on all of my devices (Androids 4.1, 4.2, 4.4, 5.1, 6.0 and 8.1) with wifi and cellular connections. I have encountered situations when my first try with wifi only and finalize form offline failed on Androids 4.4 and 5.1 but more randomly than in reproducible way and after one fail in the second attemp everything worked well and both submissions were autosent.
Surprisingly, I had no problems like that and any other on Androids 4.1 and 4.2.
I have found a case which is failing on that branch: autosend only wifi on, finalized form offline, turn on cellular data, turn off cellular data and turn on wifi - form is not autosent. I've checked also an opposite case and switched wifi and cellular data and I got same result - form wasn't send. It was reproducible on all my devices. This case succeeded on master so it might be some kind of regression. 

> The next priority is to see whether this works more broadly on Android 5.0+. Does autosend get triggered when Collect is in the background? After a reboot? After a force kill of Collect? Any other creative conditions you can think of? For now, please use Android 6 because I don't have a device that runs it and it is a popular version (21% of users). I expect other versions 5.0+ will work the same way so let's get that information first to see whether this is a direction worth pursuing and then we can do a deeper pass after some cleanup.

I performed some deeper verification on Androids 6.0 and 8.1. Nothing suspecious occured to me. I tried different cases with app in the background, reboot devices and killing collect and comibne it also with turning on internet connection (wifi/cellular) in different moments. 

@lognaturel, I hope that I understood everything in correct way :). If there is more testing needed or I missed something improtant please let me know.

jd-alexander commented:
> @jd-alexander Since this is blocking for the target API increase, I can take it from here to minimize the back-and-forth and will create a new PR tomorrow. Thanks for your work on it so far. If you have any insights on expected behavior on 4.x from your research or your own manual verification, that would be much appreciated.

No problem. I am going to be taking another look at it. I know that on 4.x it defaults to other components so I would have to do a debugging session to figure out what's going on. Been putting some work into the hierarchy view arch. so got a bit distracted trying to make some progress there so I didn't do all the proper checks here. Learning how to mutitask and be more thorough better as we go along 😄

lognaturel commented:
Continuing with narrower PR at #2619.
